development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,305 @@
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
+
+The ``$images`` array would look something like (as you can see the image paths are
+contained in the array keys):
+
+.. code-block:: php
+
+  array(
+      'ext/acme/demo/images/image1.png' => 'demo',
+      'ext/acme/demo/images/image2.png' => 'demo',
+  );
+
+
+Using service replacement
+=========================
+.. warning::
+  Only use service replacements if your extensions needs to be compatible with
+  phpBB 3.1. For 3.2 and later, see `Using service decoration`_.
+
+.. warning::
+  You can't assume the order of a service is defined in phpBB if other extensions
+  are installed. Be really carefully when using service replacements.
+
+With a service replacement you can replace a existing service in phpBB (Or in an
+extension) with your own service. Your replacement should type match the original
+service (so if the original service implements a interface, you should minimal
+implement that specific interface, if it is a concrete class, you will need to
+extent that class). The best way to accomplish this is by extending the original
+class, and only replace the features you want to change.
+
+To replace a core phpBB service, you simply name your extensions service name
+with the same name as the service in phpBB's core.
+
+If you want, for example, replace the ``config`` service in phpBB with your own
+implementation, your service configuration will look like this:
+
+.. code-block:: yaml
+
+  config:
+      class: acme\demo\config\db
+      arguments:
+          - '@dbal.conn'
+          - '@cache.driver'
+          - '%tables.config%'
+          - '@acme.demo.db_reader'
+
+Your php class should extends, at least, ``\phpbb\config\config``, which is the base
+class phpBB extends from. However, to be sure everything still works, you should
+extend from ``\phpbb\config\db``, which is the original class the service had.
+
+.. warning::
+  If you are using EPV in travis, or during submission toe the extensions database
+  at phpBB.com, you will receive a warning that your service configuration
+  doesn't follow the extensions database policies. As you are overwriting a core
+  service, you can simply ignore this message. However, in all cases you should
+  inform the the phpBB extensions team why you got the warning.
+
+Using service decoration
+========================
+.. note::
+  See for full documentation regarding service decoration the
+  `Symfony <http://symfony.com/doc/current/components/dependency_injection/advanced.html#decorating-services>`_
+  documentation.
+
+Instead of replacing a service, you should decorate the service. The will
```

Comment:
From phpBB 3.2, you can use service decoration as the preferred method to replace existing services, in the core or from other extensions. Decoration will update an existing service with a new name, leaving it intact so that it can be referenced in the new service.